Shared memory doc <master> [8054] (#103)
eProsima/Fast-RTPS-docs: docs/advanced.rst

|    :end-before: <!--><-->                     |
+-----------------------------------------------+

.. _comm-transports-shm:

Shared memory Transport (SHM)
=============================

Shared memory transport enables fast communications between entities running in the same processing unit / machine,

> these communications are based on the shared memory mechanisms provided by the host operating system.

Reviewer: ```suggestion
The shared memory transport enables fast communications between entities running in the same processing unit/machine, relying on the shared memory mechanisms provided by the host operating system.
```

SHM transport provides better performance compared with other transports like UDP / TCP, even when these transports use

> loopback interface. This is mainly due to the following reasons:

Reviewer: New line after period
```suggestion
SHM transport provides better performance than other transports like UDP / TCP, even when these transports use loopback interface.
This is mainly due to the following reasons:
```

 * Large message support: Network protocols need to fragment data in order to comply with the specific protocol and
   network stacks requirements. SHM transport allows the copy of full messages where the only size limit is the
   machine's memory capacity.

 * Reduce the number of memory copies: When sending the same message to different endpoints, SHM transport can
   directly share the same memory buffer with all the destination endpoints. Other protocols require to perform a
   copy of the message per every endpoint.

 * Less operating system overhead: Once initial setup is completed, shared memory transfers require much less system

>    calls than the other protocols. Therefore there is a performance / time consume gain by using SHM.

Reviewer: ```suggestion
 * Large message support: Network protocols need to fragment data in order to comply with the specific protocol and
   network stacks requirements.
   SHM transport allows the copy of full messages where the only size limit is the machine's memory capacity.

 * Reduce the number of memory copies: When sending the same message to different endpoints, SHM transport can
   directly share the same memory buffer with all the destination endpoints.
   Other protocols require to perform one copy of the message per endpoint.

 * Less operating system overhead: Once initial setup is completed, shared memory transfers require much less system
   calls than the other protocols.
   Therefore there is a performance/time consume gain by using SHM.
```

When two participants has SHM transport enabled and belongs to the same machine, all communications between them are
automatically performed only by SHM transport, the rest of the enabled transports are not used between those two

> participants.

Reviewer: ```suggestion
When two participants on the same machine have SHM transport enabled, all communications between them are automatically
performed by SHM transport only.
The rest of the enabled transports are not used between those two participants.
```

   references. This parameter specifies the queue size (in messages).

 * <port_overflow_policy>: Values either DISCARD or FAIL. Indicates what to do when the Listener queue is full. DISCARD
   drop the message without notifying any error, FAIL drop the message but throws an error.

 * <segment_overflow_policy>: Values either DISCARD or FAIL. Indicates what to do when there is no available space
   for messages in the shared memory segment.

 * <healthy_check_timeout_ms>: Readers and writers use a queue in shared memory to pass the messages (called Port). If
   one of the processes involved crashes while using the port, the structure can be left inoperative. So every time
   a port is opened a healthy check is performed, if the attached processes doesn't respond in
   <healthy_check_timeout_ms> milliseconds, the port is destroyed and created again.

 * <rtps_dump_file>: Full path, including the file name, of the protocol dump_file. When this string parameter is not
   empty, all the participant SHM traffic (sent and received) is traced to a file. The file format is tcpdump test hex,

>    and can be read with protocol analyzer applications.

Reviewer: I very much like this. You'll have to add tcpdump to the dictionary.
```suggestion
 * ``segment_size``: The size of the shared memory segment in bytes.
   A shared memory segment is created by each participant.
   Participant's writers copy their messages into the segment and send a message reference to the destination readers.

 * ``port_queue_capacity``: Each participant with SHM transport enabled listens on a queue (port) for incoming SHM
   message references.
   This parameter specifies the queue size (in messages).

 * ``port_overflow_policy``: Possible values are either :class:`DISCARD` or :class:`FAIL`.
   Indicates what to do when the Listener queue is full.
   :class:`DISCARD` drops the message without notifying any error, :class:`FAIL` drops the message but throws an error.

 * ``segment_overflow_policy``: Possible values are either :class:`DISCARD` or :class:`FAIL`.
   Indicates what to do when there is no available space for messages in the shared memory segment.

 * ``healthy_check_timeout_ms``: With SHM, Readers and writers use a queue to exchange messages (called Port).
   If one of the processes involved crashes while using the port, the structure can be left inoperative.
   For this reason, every time a port is opened, a healthy check is performed.
   If the attached processes doesn't respond in ``healthy_check_timeout_ms`` milliseconds, the port is destroyed and
   created again.

 * ``rtps_dump_file``: Full path, including the file name, of the protocol dump_file.
   When this string parameter is not empty, all the participant's SHM traffic (sent and received) is traced to a file.
   The file format is *tcpdump* test hex, and can be read with protocol analyzer applications such as Wireshark.
```
